File: src/javaasync/message/Message.java
package javaasync.message;

import javaasync.Employee;

/**
 * Employees are exchanging information through Messages.
 */
public class Message extends MessageBase
{

    /*
     * The message content.
     */
    private final Object content;

    /*
     * Check if message is used.
     */
    private boolean used;

    /**
     * Constructor.
     *
     * @param sender - sender.
     * @param content - message content.
     */
    public Message(Employee sender, Object content)
    {
        super(sender);
        this.content = content;
        used = false;
        type = content.getClass().getSimpleName();
    }

    /**
     * Get message content.
     *
     * @param <T> - message type.
     * @return message content.
     */
    @SuppressWarnings("unchecked")
    public <T> T content()
    {
        used = true;
        return (T) content;
    }

    /**
     * Check if message is used.
     *
     * @return true if used, otherwise false.
     */
    @Override
    public boolean used()
    {
        return used;
    }
}

File: src/javaasync/message/Message.kt
package javaasync.message

import javaasync.AsyncService

/**
 * Message base.
 */
abstract class Message(protected val sender: AsyncService) {
    /*
     * Message owner. On message creation, message sender is message owner.
     * After message sending, message receiver becomes message owner.
     */
    private var owner: AsyncService

    /*
     * Message type.
     */
    protected var type: String? = null

    init {
        this.owner = sender
        type = null
    }

    /**
     * Set receiver as owner.
     *
     * @param asyncService - receiver to set as owner.
     */
    fun setOwner(asyncService: AsyncService) {
        owner = asyncService
    }

    /**
     * Get message type.
     *
     * @return string that represents message type.
     */
    fun type(): String? {
        return type
    }

    /**
     * Get owner name.
     *
     * @return owner name.
     */
    fun owner(): AsyncService? {
        return owner
    }

    /**
     * Check if employee have access to this.
     *
     * @param asyncService - employee to check.
     * @return true if employee have access, otherwise false.
     */
    fun access(asyncService: AsyncService): Boolean {
        return asyncService.name() == owner.name()
    }

    /**
     * Get sender.
     *
     * @return sender.
     */
    fun sender(): AsyncService {
        return sender
    }

    /**
     * Check if message is used.
     *
     * @return true if used, otherwise false.
     */
    abstract fun used(): Boolean
}
